### PR TITLE
[sc 455] improve match linearization pass

### PR DESCRIPTION
**Pull request by LunaAmora**

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -155,8 +159,8 @@ pub fn desugar_book(book: &mut Book, opts: CompileOpts) -> Result<Vec<Warning>,
   if opts.eta {
     ctx.book.eta_reduction();
   }
-  if opts.supercombinators {
-    ctx.book.detach_supercombinators();
+  if opts.lift_combinators {
+    ctx.book.lift_combinators(matches!(opts.linearize_matches, OptLevel::Extra));
```
Comment (imaqtkatt) [inline src/lib.rs:163]: I think that it's better to change this `matches!` to an OptLevel method
`ctx.book.lift_combinators(opts.linearize_matches.is_extra())`

**Review by imaqtkatt — APPROVED** [review]

File: src/lib.rs [hunk 2]
```diff
@@ -387,23 +413,26 @@ impl CompileOpts {
...
   }
 }
 
 impl CompileOpts {
   pub fn check(&self, lazy_mode: bool) {
-    if !self.supercombinators && !lazy_mode {
+    if !self.lift_combinators && !lazy_mode {
       println!(
-        "Warning: Running in strict mode without enabling the supercombinators pass can lead to some functions expanding infinitely."
+        "Warning: Running in strict mode without enabling the lift_combinators pass can lead to some functions expanding infinitely."
```
Comment (developedby) [inline src/lib.rs:435]: Add a similar comment for match linearization

File: docs/compiler-options.md [hunk 3]
```diff
@@ -8,7 +8,7 @@
 | `-Oref-to-ref` `-Ono-ref-to-ref` | Disabled | [ref-to-ref](#ref-to-ref) |
 | `-Oprune` `-Ono-prune` | Disabled | [definition-pruning](#definition-pruning) |
 | `-Opre-reduce` `-Ono-pre-reduce` | Disabled | [pre-reduce](#pre-reduce) |
-| `-Osupercombinators` `-Ono-supercombinators` | Enabled  | [supercombinators](#supercombinators) |
+| `-Olift_combinators` `-Ono-lift_combinators` | Enabled  | [lift-combinators](#lift-combinators) |
```
Comment (developedby) [inline docs/compiler-options.md:11]: Can you add documentation on linearize-matches?

File: src/term/transform/linearize_matches.rs [hunk 4]
```diff
@@ -5,16 +5,16 @@ use crate::{
 };
 use indexmap::IndexSet;
 use itertools::Itertools;
-use std::collections::BTreeSet;
+use std::collections::{BTreeMap, BTreeSet};
 
 impl Ctx<'_> {
   /// Linearizes the variables between match cases, transforming them into combinators when possible.
-  pub fn linearize_simple_matches(&mut self) -> Result<(), Info> {
+  pub fn linearize_matches(&mut self, linearize_extra: bool) -> Result<(), Info> {
```
Comment (developedby) [inline src/term/transform/linearize_matches.rs:12]: linearize_extra is not very descriptive

**Review by developedby — CHANGES_REQUESTED** [review]

File: docs/compiler-options.md [hunk 5]
```diff
@@ -134,7 +135,30 @@ main = (bar)
 @main = @bar
 ```
 
-## Supercombinators
+## linearize-matches
+
+Linearizes the variables between match cases, transforming them into combinators when possible.  
+When `linearize-matches` linearizes only vars that are used in more than one arm.
```
Comment (developedby) [inline docs/compiler-options.md:141]: When the `linearize-matches` option is used, ...

File: docs/compiler-options.md [hunk 6]
```diff
@@ -134,7 +135,30 @@ main = (bar)
...
+
+Example:
+```rs
+λa λb match a { 0: b; +: b }
+
+// Is tranformed to
+λa λb (match a { 0: λc c; +: λd d } b)
+```
+
+When `linearize-matches-extra`, linearizes all vars used in the arms.
```
Comment (developedby) [inline docs/compiler-options.md:151]: Same here

**Review by developedby — CHANGES_REQUESTED** [review]

**Review by developedby — APPROVED** [review]

**Pull request merged**